Transfer log events as JSON objects and allow filtering on front-end

walterl opened the pull request:
The only visible changes to the UI is the added process `<select>`, and the way in which the log events are rendered. The event rendering could probably be a bit nicer.

![image](https://user-images.githubusercontent.com/23798/62655593-9dfc4500-b962-11e9-889d-c127e903af7d.png)

![image](https://user-images.githubusercontent.com/23798/62655694-d439c480-b962-11e9-8039-7985aa4e7afe.png)

Should the select perhaps be disabled with the available UUID selected, when there is only a single process UUID available? Or maybe auto-selecting _All Metabase processes_ in that case is better...?

walterl committed:
Return log events as JSON objects from API; render on front-end

Site UUID added to log events.

walterl committed:
Remove superfluous memoization of site UUID retrieval

Settings are already cached.

walterl committed:
Transfer timestamps in ISO-8601 format

walterl committed:
Move `metabase.metabot.instance/local-process-uuid` to `metabase.public-settings`

walterl committed:
Use `local-process-uuid` in stead of `site-uuid`

`site-uuid` is not unique across processes, which is what we want to
associate with log events.

walterl committed:
Remove unused import

walterl committed:
Allow log events to be filtered by process UUID on front-end

walterl committed:
Move process selector out of loading wrapper

Otherwise, if there are process UUIDs with no log events, the selector is
removed with the log event list.

walterl committed:
Figured out how to add docs to `defonce`

walterl committed:
Fix ns declaration

walterl committed:
Linter appeasement

walterl commented:
This fixes #9846.

walterl committed:
Merge log events received from back-end with previously received ones

This allows responses from multiple back-end instances to be spliced
together.

walterl committed:
Use MB components in stead of plain, styled HTML tags

The select is only displayed if more than one process UUID was found.

walterl committed:
Linter appeasement

camsaul commented on src/metabase/public_settings.clj:
```diff
@@ -232,3 +232,10 @@
+                run. Restarting the server will change this UUID, and each server in a hortizontal cluster will have its own ID,
+                making this different from the `site-uuid` Setting. The local process UUID is used to differentiate different
+                horizontally clustered MB instances so we can determine which of them will handle MetaBot duties."}
+  local-process-uuid (str (UUID/randomUUID)))
```
Sorry, I'm a knucklehead. I thought for some reason `local-process-uuid` was a Setting the same way `site-uuid` is. Since this is a plain UUID I think we should go ahead and put it in `metabase.config` instead. Sorry about that.

walterl replied:
Done :wrench:

camsaul reviewed (approved):
Backend stuff looks good once you move `site-uuid` again (sorry 😓), pinging @tlrobinson to check out FE changes

walterl committed:
Move `metabase.public-settings/local-process-uuid` to `metabase.config`

tlrobinson commented on frontend/src/metabase/admin/tasks/containers/Logs.jsx:
```diff
@@ -80,23 +94,73 @@ export default class Logs extends Component {
+
+    let processUUIDSelect = null;
+    if (processUUIDs.length > 1) {
+      const options = [
```
You can just do something like this if you have a mix of static and dynamic options: 

```
<Select ...>
  <Option value="ALL" key="ALL">{t`All Metabase processes`}</Option>
  { processUUIDs.map(uuid => (
    <Option key={uuid} value={uuid}>
      <code>{uuid}</code>
    </Option>
  ))}
</Select>

walterl replied:
:heavy_check_mark: 

There was a bug in `Select` that prevented that. But I fixed it now :)

tlrobinson reviewed (changes requested):
Looks good except a few little things

walterl commented on frontend/src/metabase/admin/tasks/containers/Logs.jsx:
```diff
@@ -28,12 +31,23 @@ for (const [name, color] of Object.entries(ANSI_COLORS)) {
+
+function mergeLogs(...logArrays) {
+  let logs = Array.prototype.concat(...logArrays);
+  logs = _.sortBy(logs, ev => [ev.timestamp, ev.process_uuid, ev.msg]);
```
Thanks for that. My testing was clearly insufficient. :+1: 

I'm sorting by `msg` too, just to make sure that the output is stable. :)

walterl replied:
:heavy_check_mark: 

All 3 of those fields (`timestamp`, `process_uuid` and `msg`) are strings, so sorting by the string representation of the array would've been the same :stuck_out_tongue_closed_eyes:

walterl committed:
Fix sorting on multiple log event fields

walterl committed:
Flatten `Select` children to allow nested arrays of children

The allows the mixing of static and dynamicly generated children:

```
<Select ...>
  <Option value="static">Static</Option>
  {options.map(o => <Option value={o}>Dynamic {o}</Option>}
</Select>

walterl committed:
Generate options directly under parent Select

walterl committed:
Limit the number of log events that we render

walterl commented on frontend/src/metabase/admin/tasks/containers/Logs.jsx:
```diff
@@ -28,12 +31,23 @@ for (const [name, color] of Object.entries(ANSI_COLORS)) {
+}
+
+function mergeLogs(...logArrays) {
+  let logs = Array.prototype.concat(...logArrays);
```
:heavy_check_mark: and :heavy_check_mark:

walterl commented on frontend/src/metabase/admin/tasks/containers/Logs.jsx:
```diff
@@ -28,12 +31,23 @@ for (const [name, color] of Object.entries(ANSI_COLORS)) {
   addCSSRule(`.react-ansi-style-${name}`, `color: ${color} !important`);
 }
 
+function logEventUniqueValue(ev) {
```
:heavy_check_mark:

walterl commented:
@tlrobinson Besides the code, I'm not quite happy with the UI itself, and would appreciate some feedback:

![image](https://user-images.githubusercontent.com/23798/62805734-2a3c7280-baf1-11e9-8340-b2965819d65b.png)

I couldn't figure out how to fix the wrapping in the menu, or the full width of the select, with only the CSS in `metabase/css/core`.

camsaul commented on src/metabase/config.clj:
```diff
@@ -101,6 +102,13 @@
+(defonce ^{:doc "This UUID is randomly-generated upon launch and used to identify this specific Metabase instance during this specifc
+                run. Restarting the server will change this UUID, and each server in a hortizontal cluster will have its own ID,
+                making this different from the `site-uuid` Setting. The local process UUID is used to differentiate different
+                horizontally clustered MB instances so we can determine which of them will handle MetaBot duties."}
```
Since this is not MetaBot specific anymore please update the docstring 😉

walterl replied:
:heavy_check_mark:

camsaul commented on src/metabase/config.clj:
```diff
@@ -101,6 +102,13 @@
+                run. Restarting the server will change this UUID, and each server in a hortizontal cluster will have its own ID,
+                making this different from the `site-uuid` Setting. The local process UUID is used to differentiate different
+                horizontally clustered MB instances so we can determine which of them will handle MetaBot duties."}
+  local-process-uuid (str (UUID/randomUUID)))
```
A couple other style things:

*  Since this is a multiline `def` form you should put `(str (UUID/randomUUID))` on its own line
*  Our general column width goal is 120 chars, because that's what GitHub shows without cutting it off. Tweak your editor to do that for you. If you want to add a config file for your editor to the repo feel free to do so, we already have one that has those sorts of settings for Emacs. 
*  I don't think it's worth putting a comment about where this variable used to live, hopefully everyone can forget it ever lived there moving forward. It's not info that affects how you'd use it so let's not clutter people's heads with that info, you can always go in Git history if you want to see the history of the code base

walterl replied:
Thanks for that! :+1:

walterl committed:
Docstring and style updates

walterl committed:
Update editor config with Clojure files max line length

walterl committed:
Ensure that `NumericWrapper` is imported before use

tlrobinson committed:
Cleanup layout
